Enable extension pausing when in "Recovery Mode"

TimothyBJacobs opened the pull request:
See [#46130](https://core.trac.wordpress.org/ticket/46130#comment:6) for an in-depth explanation.

Commit:
Re-introduce original fatal error recovery mechanism.

TimothyBJacobs committed:
Migrate patch 46130.1.diff

felixarntz commented on src/wp-admin/includes/class-wp-plugins-list-table.php:
```diff
@@ -911,11 +904,18 @@ public function single_row( $item ) {
 
 							$error['type'] = $core_errors[ $error['type'] ];
 
+							if ( empty( $error['wp_is_protected'] ) ) {
```
I don't think we need to differentiate here. I think only errors that happen during recovery mode should be stored, and only the latest one per extension.

TimothyBJacobs replied:
This is no longer differentiated it. We store all errors in recovery mode and only email for errors in protected endpoints.

felixarntz commented on src/wp-includes/class-wp-fatal-error-handler.php:
```diff
@@ -40,10 +40,12 @@ public function handle() {
 			// If the error was stored and thus the extension paused,
 			// redirect the request to catch multiple errors in one go.
-			if ( $this->store_error( $error ) ) {
+			if ( $this->store_error( $error ) && wp_is_recovery_mode() ) {
```
I think the two conditions should be ordered the other way around, if we only store errors in recovery mode.

felixarntz commented on src/wp-includes/error-protection.php:
```diff
@@ -88,6 +88,8 @@ function wp_record_extension_error( $error ) {
 	$parts     = explode( '/', $path );
 	$extension = array_shift( $parts );
 
+	$error['wp_is_protected'] = is_protected_endpoint();
```
Per the above, this shouldn't be necessary.

felixarntz commented on src/wp-includes/pluggable.php:
```diff
@@ -979,6 +979,8 @@ function wp_clear_auth_cookie() {
 		// Post password cookie
 		setcookie( 'wp-postpass_' . COOKIEHASH, ' ', time() - YEAR_IN_SECONDS, COOKIEPATH, COOKIE_DOMAIN );
+
+		clear_recovery_mode_cookie();
```
Can we maybe hook into an action instead? This could be problematic since the function might be overridden. On the other hand, we might not even want to clear that cookie here, as it is not tied to a logged-in user anyway.

TimothyBJacobs replied:
Can do. I lean towards clearing it since recovery mode is something security sensitive, even if it isn't actually authenticating the user.

felixarntz commented on src/wp-login.php:
```diff
@@ -438,7 +438,7 @@ function retrieve_password() {
 
 // Validate action so as to default to the login screen.
-if ( ! in_array( $action, array( 'postpass', 'logout', 'lostpassword', 'retrievepassword', 'resetpass', 'rp', 'register', 'login', 'confirmaction' ), true ) && false === has_filter( 'login_form_' . $action ) ) {
+if ( ! in_array( $action, array( 'postpass', 'logout', 'lostpassword', 'retrievepassword', 'resetpass', 'rp', 'register', 'login', 'confirmaction', 'begun_recovery_mode' ), true ) && false === has_filter( 'login_form_' . $action ) ) {
```
I suggest using `entered_recovery_mode` instead.

felixarntz commented on src/wp-includes/load.php:
```diff
@@ -807,6 +807,483 @@ function wp_skip_paused_themes( array $themes ) {
+/**
+ * Create a recovery mode key for a user.
+ *
+ * @since 5.1.0
```
Version numbers can be updated to 5.2.0. :)

felixarntz commented on src/wp-includes/load.php:
```diff
@@ -807,6 +807,483 @@ function wp_skip_paused_themes( array $themes ) {
+	global $wp_hasher;
+
+	if ( ! function_exists( 'wp_generate_password' ) ) {
+		require_once ABSPATH . WPINC . '/pluggable.php';
```
Such inclusions should happen in the same function that also dies at the end of its execution, to make sure it never happens at a point when it shouldn't.

I generally suggest adding a class for the recovery mode functionality since it's quite complex and several functions would better be private/protected.

TimothyBJacobs replied:
Agreed.

TimothyBJacobs replied:
Moved to a class.

felixarntz commented on src/wp-includes/load.php:
```diff
@@ -807,6 +807,483 @@ function wp_skip_paused_themes( array $themes ) {
+	 *
+	 * @param string  $key  The recovery mode key.
+	 */
+	do_action( 'generate_recovery_mode_key', $key );
```
Do we need this action?

TimothyBJacobs replied:
We should probably do a pass on filters/actions after everything else is ready. I think extension pieces are important here for hosts that might want to handle things differently. I think its probably too complex for replacing wholesale via a drop-in.

felixarntz commented on src/wp-includes/load.php:
```diff
@@ -807,6 +807,483 @@ function wp_skip_paused_themes( array $themes ) {
+
+	$hashed = $wp_hasher->HashPassword( $key );
+
+	update_site_option( 'recovery_key', array(
```
What was the reasoning behind using `*_site_option()` instead of `*_option()`?

TimothyBJacobs replied:
This matches the behavior in `wp_salt()`

felixarntz commented on src/wp-includes/load.php:
```diff
@@ -807,6 +807,483 @@ function wp_skip_paused_themes( array $themes ) {
+ */
+function validate_recovery_mode_key( $key ) {
+
+	$record = get_site_option( 'recovery_key' );
```
See above.

felixarntz commented on src/wp-includes/load.php:
```diff
@@ -807,6 +807,483 @@ function wp_skip_paused_themes( array $themes ) {
+	}
+
+	if ( ! function_exists( 'wp_check_password' ) ) {
+		require_once ABSPATH . WPINC . '/pluggable.php';
```
See above, such an inclusion shouldn't happen in a function that does not terminate the request and is publicly callable.

felixarntz commented on src/wp-includes/load.php:
```diff
@@ -807,6 +807,483 @@ function wp_skip_paused_themes( array $themes ) {
+ */
+function recovery_mode_hash( $data ) {
+
+	if ( ! defined( 'AUTH_KEY' ) || AUTH_KEY === 'put your unique phrase here' ) {
```
Use Yoda condition you must. 😛

felixarntz commented on src/wp-includes/load.php:
```diff
@@ -807,6 +807,483 @@ function wp_skip_paused_themes( array $themes ) {
+
+		if ( ! $auth_key ) {
+			if ( ! function_exists( 'wp_generate_password' ) ) {
+				require_once ABSPATH . WPINC . '/pluggable.php';
```
See above.

felixarntz commented on src/wp-includes/load.php:
```diff
@@ -807,6 +807,483 @@ function wp_skip_paused_themes( array $themes ) {
+
+		if ( ! $auth_salt ) {
+			if ( ! function_exists( 'wp_generate_password' ) ) {
+				require_once ABSPATH . WPINC . '/pluggable.php';
```
See above.

felixarntz commented on src/wp-includes/load.php:
```diff
@@ -807,6 +807,483 @@ function wp_skip_paused_themes( array $themes ) {
+function generate_recovery_mode_cookie() {
+
+	if ( ! function_exists( 'wp_generate_password' ) ) {
+		require_once ABSPATH . WPINC . '/pluggable.php';
```
See above.

felixarntz commented on src/wp-includes/load.php:
```diff
@@ -807,6 +807,483 @@ function wp_skip_paused_themes( array $themes ) {
+
+	if ( ! $cookie ) {
+		if ( empty( $_COOKIE[ RECOVERY_MODE_COOKIE ] ) ) {
+			return new WP_Error( 'no_cookie' );
```
Missing error message.

felixarntz commented on src/wp-includes/load.php:
```diff
@@ -807,6 +807,483 @@ function wp_skip_paused_themes( array $themes ) {
+		return;
+	}
+
+	if ( isset( $_GET['action'], $_GET['rm_key'] ) && 'begin_recovery_mode' === $_GET['action'] ) {
```
I suggest using `enter_recovery_mode`.

TimothyBJacobs replied:
Works for me

felixarntz commented on src/wp-includes/load.php:
```diff
@@ -807,6 +807,483 @@ function wp_skip_paused_themes( array $themes ) {
+		}
+
+		$url     = add_query_arg( 'action', 'begun_recovery_mode', wp_login_url() );
+		$message = '<script type="application/javascript">window.location = \'' . esc_js( $url ) . '\';</script>';
```
I read your comment on why you did that, but I think we should further look into why redirecting via PHP doesn't work. The approach here seems weird.

TimothyBJacobs replied:
Yeah, I'm not sure what was going on there. Setting the cookie and redirect should work, that's how `wp-login.php` happens. So I was probably seeing some other behavior.

felixarntz commented on src/wp-includes/load.php:
```diff
@@ -807,6 +807,483 @@ function wp_skip_paused_themes( array $themes ) {
+	$url = add_query_arg( array(
+		'action' => 'begin_recovery_mode',
+		'rm_key' => $key,
+	), wp_login_url() );
```
This violates WP Coding Standards :)

TimothyBJacobs replied:
Sorry, how are we supposed to format that now?

swissspidy replied:
@TimothyBJacobs if a function call is multi-line, each parameter should start on a new line:

```php
$url = add_query_arg(
	array(
		'action' => 'begin_recovery_mode',
		'rm_key' => $key,
	),
	wp_login_url()
);
```

TimothyBJacobs replied:
Thank you @swissspidy! will update

felixarntz reviewed (commented):
Great work so far! I pointed out a few concerns. Here are some more general observations and thoughts (partly also touched in inline comments):

* I think there should be a class for recovery mode functionality. `WP_Recovery_Mode` or something like that (should probably be a `final` class) - simply for the ability to better group functionality together and to make methods private that shouldn't be public.
* No files should be included anywhere but in code that a 100% surely terminates the request itself.
* We need to think about whether to use options or network options (in case we're in a multisite that's relevant). I think network options certainly make sense for the fallback key and salt that just replace a constant - for the rest, that's worth discussing.
* I don't think we need the concept of protected endpoints any longer. If in recovery mode, pause extensions - if not, don't. IMO there's no need for this additional complexity, i.e. the `is_protected_endpoint()` should be dropped completely in favor of `wp_is_recovery_mode()`.

+1 on using custom code for the hashing etc. - since we can't use the pluggable functions in these circumstances, that's a fine approach I'd say.

swissspidy commented on src/wp-admin/includes/class-wp-plugins-list-table.php:
```diff
@@ -911,11 +904,18 @@ public function single_row( $item ) {
+								$error_message = __( 'The plugin caused an error of type %1$s in line %2$s of the file %3$s. Error message: %4$s' );
+							} else {
+								/* translators: 1: error type, 2: error line number, 3: error file name, 4: error message */
+								$error_message = __( 'The plugin caused an error in the <strong>admin backend</strong> of type %1$s in line %2$s of the file %3$s. Error message: %4$s' );
```
FYI in WordPress we use the noun `back end` instead of `backend`, see https://make.wordpress.org/core/handbook/best-practices/spelling/.

Also, `admin backend` is a bit of a pleonasm. I'd either use `admin` or `back end`, not both.

TimothyBJacobs replied:
Updated

TimothyBJacobs committed:
Refactor of recovery mode to encapsulate behavior in a class.

This introduces a WP_Recovery_Mode_Controller interface. Core would ship
with the email controller which handles recovery mode by sending an
email and setting a cookie. But this will allow for hosts or mu-plugins
to substitute any handler that implements that interface.

The cookie handler has also been abstracted out to a separate service
and can be reused by a custom handler. It also allows for substituting
all the cookie config values in the constructor. This should make it easier
to ship this eventually to network activated plugins as well, since a primary
concern there is that the correct cookie constants have not been populated
until after those plugins have been loaded.

This does not yet drop `is_protected_endpoint()` since that is still to be
discussed. Additionally, I have not had time yet to refactor the extension
storage to use one option as @flixos90 suggested.

TimothyBJacobs committed:
Use enter_recovery_mode instead of begin_recovery_mode.

TimothyBJacobs committed:
Extract checking if the recovery cookie is set at all to the cookie service.

TimothyBJacobs committed:
- Add more details to recovery mode email.
- Support RECOVERY_MODE_EMAIL constant to set the email address.
- Change default rate limit to 4 hours.
- Ensure link is valid for at least as long as the rate limit duration.
- Only send recovery mode email when the errors occurs in a protected endpoint.

TimothyBJacobs committed:
Consolidate paused extension storage to a single option.

TimothyBJacobs committed:
Actually pull email address from method.

TimothyBJacobs committed:
Split email key into a separate service class.

This allows for better reusability for alternate recovery mode controllers
and allows for better testing.

TimothyBJacobs committed:
Allow overriding the recovery mode controller with a drop-in and a filter.

TimothyBJacobs committed:
Add a method on WP_Recovery_Mode_Controller to handle the error instead
of firing an action.

This mostly reflects the Google Docs proposed version. I don't think the
recovery mode controller should at all be responsible for pausing extensions.
This makes extending the functionality through drop-ins unsafe since we are
moving unrelated functionality into that method. If we change how extensions
are recorded or want to perform different behavior, the drop-ins would be
incompatible.

TimothyBJacobs committed:
Fix unintentional re-ordering of block includes.

TimothyBJacobs commented:
User Flow & Technical Doc: https://docs.google.com/document/d/1XWQghXj0p0X8GF-Byuv4vfPQG0Rs0XoVfOepVjiMkfM/edit
From Slack: 

I do think we are in a good place extensibility wise to be able to support a user email specific flow in 5.3 with our current framework and IMO could be implemented as a drop-in with some minor changes. The main thing being an `exit_recovery_mode()` method of some kind.

I think that putting an exit mode in the admin bar makes sense. I think it’d be quickest to add the underlying technical support for exiting recovery mode, like a method, in the patch before commit and then worry about that specific UI shortly thereafter.
From a thread in #core-phpToday at 8:55 AMView reply

Looking at a specific of the technical document, I don’t think we should overload the `is_active()` method on the controller. That will make it harder to support more advanced recovery mode procedures in the future. I’m fairly confident that just having a session ID will be enough to support per-user recovery mode, ie we don’t actually need to provide the user ID. If we do find that we need actual data about the user, then I think that’d be better off by implementing a specific `WP_User_Specific_Recovery_Mode_Controller` interface with a `get_user_id(): int` method ( or something similar ) that way we can properly handle recovery mode controllers that are site wide, like the current controller.
I also think we should keep the separation of concerns with the additional helper classes, instead of moving many of these to private methods. Moving them into separate classes allows for better unit testability, see `Tests_Recovery_Mode`, eases understanding by grouping similar functionality together, and makes it substantially easier for drop-ins or other customer recovery mode controllers to work. This will give them the helper libraries they need so they aren’t forced to reinvent the wheel or copy parts of the original controller implementation .

felixarntz commented:
> Looking at a specific of the technical document, I don’t think we should overload the `is_active()` method on the controller. That will make it harder to support more advanced recovery mode procedures in the future. I’m fairly confident that just having a session ID will be enough to support per-user recovery mode, ie we don’t actually need to provide the user ID.

Given the current approach, this completely makes sense. Let's have a `is_active()` and separate `get_session_id()` method. Later, we could potentially store a session ID in user meta to associate them, if needed.

> I also think we should keep the separation of concerns with the additional helper classes, instead of moving many of these to private methods.

Yes. The reason I made them private methods in the doc was mostly to have a better understanding when reading about them - I see it as a starting point to abstract from further. The multiple classes approach is better organized, but a bit harder to grasp at first glance.

nerrad commented:
For first iteration (release in 5.2) I think this should be disabled for WordPress multisite.  There's ux/ui workflows that need to be considered in a multisite environment and typically those working with multisite have a better handle on dealing with php/plugin upgrades etc. so the urgency for these protections in that environment isn't as great.

felixarntz commented on src/wp-admin/includes/class-wp-plugins-list-table.php:
```diff
@@ -209,6 +210,9 @@ public function prepare_items() {
 					$plugins['active'][ $plugin_file ] = $plugin_data;
+					if ( is_plugin_paused( $plugin_file ) ) {
+						$plugins['paused'][ $plugin_file ] = $plugin_data;
+					}
```
This clause should be removed, as we don't act on network-active plugins.

TimothyBJacobs replied:
Removed.

felixarntz commented on src/wp-admin/includes/class-wp-plugins-list-table.php:
```diff
@@ -218,6 +222,9 @@ public function prepare_items() {
 				$plugins['active'][ $plugin_file ] = $plugin_data;
+				if ( is_plugin_paused( $plugin_file ) ) {
+					$plugins['paused'][ $plugin_file ] = $plugin_data;
+				}
```
The check in this clause needs a `! $screen->in_admin( 'network' )` check.

TimothyBJacobs replied:
Added

felixarntz commented on src/wp-admin/includes/class-wp-plugins-list-table.php:
```diff
@@ -633,11 +644,19 @@ public function single_row( $item ) {
+					if ( current_user_can( 'manage_network_plugins' ) && count_paused_plugin_sites_for_network( $plugin_file ) ) {
+						/* translators: %s: plugin name */
+						$actions['resume'] = '<a class="resume-link" href="' . wp_nonce_url( 'plugins.php?action=resume&amp;plugin=' . urlencode( $plugin_file ) . '&amp;plugin_status=' . $context . '&amp;paged=' . $page . '&amp;s=' . $s, 'resume-plugin_' . $plugin_file ) . '" aria-label="' . esc_attr( sprintf( _x( 'Network Resume %s', 'plugin' ), $plugin_data['Name'] ) ) . '">' . __( 'Network Resume' ) . '</a>';
+					}
```
This clause should be removed.

TimothyBJacobs replied:
Removed

felixarntz commented on src/wp-admin/includes/class-wp-plugins-list-table.php:
```diff
@@ -633,11 +644,19 @@ public function single_row( $item ) {
+					if ( current_user_can( 'manage_network_plugins' ) && count_paused_plugin_sites_for_network( $plugin_file ) ) {
+						/* translators: %s: plugin name */
+						$actions['resume'] = '<a class="resume-link" href="' . wp_nonce_url( 'plugins.php?action=resume&amp;plugin=' . urlencode( $plugin_file ) . '&amp;plugin_status=' . $context . '&amp;paged=' . $page . '&amp;s=' . $s, 'resume-plugin_' . $plugin_file ) . '" aria-label="' . esc_attr( sprintf( _x( 'Network Resume %s', 'plugin' ), $plugin_data['Name'] ) ) . '">' . __( 'Network Resume' ) . '</a>';
+					}
```
This clause should be removed.

TimothyBJacobs replied:
Removed

felixarntz commented on src/wp-admin/includes/class-wp-plugins-list-table.php:
```diff
@@ -648,6 +667,10 @@ public function single_row( $item ) {
+					if ( ! $restrict_network_only && current_user_can( 'resume_plugin', $plugin_file ) && is_plugin_paused( $plugin_file ) ) {
+						/* translators: %s: plugin name */
+						$actions['resume'] = '<a class="resume-link" href="' . wp_nonce_url( 'plugins.php?action=resume&amp;plugin=' . urlencode( $plugin_file ) . '&amp;plugin_status=' . $context . '&amp;paged=' . $page . '&amp;s=' . $s, 'resume-plugin_' . $plugin_file ) . '" aria-label="' . esc_attr( sprintf( _x( 'Resume %s', 'plugin' ), $plugin_data['Name'] ) ) . '">' . __( 'Resume' ) . '</a>';
+					}
```
This clause should be removed.

TimothyBJacobs replied:
Removed

felixarntz commented on src/wp-admin/includes/class-wp-plugins-list-table.php:
```diff
@@ -764,6 +791,12 @@ public function single_row( $item ) {
 			$class .= ' update';
 		}
 
+		$paused = is_plugin_paused( $plugin_file );
```
This should first check `! $screen->in_admin( 'network' )`.

TimothyBJacobs replied:
Added a check.

felixarntz commented on src/wp-admin/includes/plugin.php:
```diff
@@ -468,12 +468,15 @@ function get_dropins() {
+		'object-cache.php'             => array( __( 'External object cache.' ), true ), // auto on load
+		'php-error.php'                => array( __( 'Custom PHP error message.' ), true ), // auto on error
+		'fatal-error-handler.php'      => array( __( 'Custom PHP fatal error handler.' ), true ), // auto on error
+		'recovery-mode-controller.php' => array( __( 'Custom recovery mode controller.' ), true ), // auto on error
```
Let's not support a `recovery-mode-controller.php` drop-in for now. This is much more complicated functionality in an early state, and furthermore it is more applicable to be modified by MU plugins. Drop-ins should only be added for things that need to run very early.

felixarntz commented on src/wp-admin/includes/plugin.php:
```diff
@@ -2101,3 +2104,198 @@ function wp_add_privacy_policy_content( $plugin_name, $policy_text ) {
+		return false;
+	}
+
+	if ( ! is_plugin_active( $plugin ) && ! is_plugin_active_for_network( $plugin ) ) {
```
The `! is_plugin_active_for_network( $plugin )` should be removed, we don't cover network-active plugins.

TimothyBJacobs replied:
Removed

felixarntz commented on src/wp-admin/includes/plugin.php:
```diff
@@ -2101,3 +2104,198 @@ function wp_add_privacy_policy_content( $plugin_name, $policy_text ) {
+ * @param string $plugin Path to the plugin file relative to the plugins directory.
+ * @return int Site count.
+ */
+function count_paused_plugin_sites_for_network( $plugin ) {
```
This function should be removed.

TimothyBJacobs replied:
Removed

felixarntz commented on src/wp-admin/includes/plugin.php:
```diff
@@ -2101,3 +2104,198 @@ function wp_add_privacy_policy_content( $plugin_name, $policy_text ) {
+		return;
+	}
+
+	if ( ! current_user_can( 'deactivate_plugins' ) ) {
```
Let's check `resume_plugins` instead.

TimothyBJacobs replied:
Changed

felixarntz commented on src/wp-admin/includes/theme.php:
```diff
@@ -768,3 +768,127 @@ function customize_themes_print_templates() {
+ * @param string $theme Path to the theme directory relative to the themes directory.
+ * @return int Site count.
+ */
+function count_paused_theme_sites_for_network( $theme ) {
```
This function should be removed.

TimothyBJacobs replied:
Removed

felixarntz commented on src/wp-admin/includes/theme.php:
```diff
@@ -768,3 +768,127 @@ function customize_themes_print_templates() {
+		return;
+	}
+
+	if ( ! current_user_can( 'switch_themes' ) ) {
```
Let's check `resume_themes` instead.

TimothyBJacobs replied:
Changed

felixarntz commented on src/wp-admin/plugins.php:
```diff
@@ -389,6 +389,27 @@
+				wp_die( __( 'Sorry, you are not allowed to resume this plugin.' ) );
+			}
+
+			if ( is_multisite() && ! is_network_admin() && is_network_only_plugin( $plugin ) ) {
```
Let's change this clause to `is_multisite() && is_network_admin()` and put it at the beginning of the `case 'resume'`, to ensure this doesn't do anything in the network admin.

TimothyBJacobs replied:
Changed.

felixarntz commented on src/wp-admin/plugins.php:
```diff
@@ -389,6 +389,27 @@
+
+			check_admin_referer( 'resume-plugin_' . $plugin );
+
+			$result = resume_plugin( $plugin, self_admin_url( 'plugins.php?error=resuming' ), is_network_admin() );
```
Do not pass the third parameter here.

TimothyBJacobs replied:
Removed.

felixarntz commented on src/wp-admin/themes.php:
```diff
@@ -33,6 +33,26 @@
+		check_admin_referer( 'resume-theme_' . $_GET['stylesheet'] );
+		$theme = wp_get_theme( $_GET['stylesheet'] );
+
+		if ( ! current_user_can( 'resume_themes' ) ) {
```
This should be `! current_user_can( 'resume_theme', $_GET['stylesheet'] )`.

TimothyBJacobs replied:
Changed

felixarntz commented on src/wp-includes/capabilities.php:
```diff
@@ -464,6 +464,9 @@ function map_meta_cap( $cap, $user_id ) {
 			break;
+		case 'resume_plugin':
+			$caps[] = 'resume_plugins';
+			break;
```
There should also be a similar clause for `resume_theme`, resolving to `resume_themes`.

TimothyBJacobs replied:
Added.

felixarntz commented on src/wp-includes/class-wp-fatal-error-handler.php:
```diff
@@ -0,0 +1,192 @@
+				}
+			} else {
+				wp_recovery_mode()->handle_error( $error );
+			}
```
Let's move this entire handling into `wp_recovery_mode()->handle_error()`, so that instead of the if-else the code only calls that method here. The method should return a boolean on whether it could successfully handle the error (false if caused by core/drop-in/MU-plugin/network-plugin, or if storing it failed due to an internal error).

TimothyBJacobs replied:
Moved.

felixarntz commented on src/wp-includes/class-wp-fatal-error-handler.php:
```diff
@@ -0,0 +1,192 @@
+		}
+
+		// Bail if this error should not be handled.
+		if ( ! wp_should_handle_error( $error ) ) {
```
Let's remove this function and move it to an internal method, since it's only used here, and we shouldn't expose more than necessary.

TimothyBJacobs replied:
Moved.

felixarntz commented on src/wp-includes/class-wp-fatal-error-handler.php:
```diff
@@ -0,0 +1,192 @@
+	 *
+	 * @return bool True if the error was stored successfully, false otherwise.
+	 */
+	protected function store_error( $error ) {
```
This should be removed from here, the recovery mode controller should take care of this.

felixarntz commented on src/wp-includes/class-wp-fatal-error-handler.php:
```diff
@@ -0,0 +1,192 @@
+	 *
+	 * @since 5.2.0
+	 */
+	protected function redirect_protected() {
```
This should be removed from here, the recovery mode controller should take care of this.

felixarntz commented on src/wp-includes/class-wp-paused-extensions-storage.php:
```diff
@@ -0,0 +1,249 @@
+	 * }
+	 * @return bool True on success, false on failure.
+	 */
+	public function record( $type, $extension, $error ) {
```
Rename this to `set()`.

TimothyBJacobs replied:
Renamed.

felixarntz commented on src/wp-includes/class-wp-paused-extensions-storage.php:
```diff
@@ -0,0 +1,249 @@
+	 * @param string $extension Plugin or theme directory name.
+	 * @return bool True on success, false on failure.
+	 */
+	public function forget( $type, $extension ) {
```
Rename this to `delete()`.

TimothyBJacobs replied:
Renamed

felixarntz commented on src/wp-includes/class-wp-paused-extensions-storage.php:
```diff
@@ -0,0 +1,249 @@
+ *
+ * @since 5.2.0
+ */
+class WP_Paused_Extensions_Storage {
```
What was the reason for changing the class from how it worked in https://github.com/wp-core-php/wordpress-develop/pull/5/files#diff-a08d55fe0e259e54b08e51ef2dea7a7c? I think it's cleaner having separate instances for plugins and themes - as they can theoretically share names. Let's have two distinct options for plugins/themes.

TimothyBJacobs replied:
You suggested consolidating it into one option :)
https://wordpress.slack.com/archives/C60K3MP2Q/p1550509401260900

TimothyBJacobs replied:
Changed into two storage instances that use the same option backend.

felixarntz commented on src/wp-includes/class-wp-paused-extensions-storage.php:
```diff
@@ -0,0 +1,249 @@
+	 * @param string $extension Plugin or theme directory name.
+	 * @return array A single clause to add to a meta query.
+	 */
+	public function get_site_meta_query_clause( $type, $extension ) {
```
Let's remove this method for now, as we don't need it (yet).

TimothyBJacobs replied:
Removed.

felixarntz commented on src/wp-includes/class-wp-paused-extensions-storage.php:
```diff
@@ -0,0 +1,249 @@
+	 *
+	 * @return string
+	 */
+	private function get_site_meta_key( $type, $extension ) {
```
Instead of having this method, both the option name and site meta key name parts that depend on whether plugin or theme should:
* either be passed via the constructor (with core taking care of this instantiation somewhere)
* or be defined in child classes (making this an abstract class and then having `WP_Paused_Plugins_Storage` and `WP_Paused_Themes_Storage` extending this)

TimothyBJacobs replied:
Dropped.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-controller.php:
```diff
@@ -0,0 +1,46 @@
+	 *
+	 * @return bool
+	 */
+	public function is_recovery_mode_active();
```
I think this should be shortened to `is_active()`.

TimothyBJacobs replied:
Shortened.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-controller.php:
```diff
@@ -0,0 +1,46 @@
+	 *
+	 * @return string|false
+	 */
+	public function get_recovery_mode_session_id();
```
I think this should be shortened to `get_session_id()`.

TimothyBJacobs replied:
Shortened

felixarntz commented on src/wp-includes/class-wp-recovery-mode-cookie-service.php:
```diff
@@ -0,0 +1,255 @@
+<?php
+
+final class WP_Recovery_Mode_Cookie_Service {
```
This class is missing quite a few docs. File, class, properties, methods should all have descriptions and `since` annotations.

TimothyBJacobs replied:
Added docs.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-controller.php:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Interface WP_Recovery_Mode_Controller
+ */
+interface WP_Recovery_Mode_Controller {
```
As much as _I_ think interfaces are a good idea, let's not do this right now. Similar to the `WP_Fatal_Error_Handler`, let's make this here a class with the default implementation (migrating the code from `WP_Recovery_Mode_Email_Controller` into here.

Introducing interfaces doesn't have a best practice yet (also in terms of coding standards, where the current `class-` prefix wouldn't make sense). It would be good to explore this in the near future, but not now for this feature, given the time constraints.

TimothyBJacobs replied:
Removed interface support.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-controller.php:
```diff
@@ -0,0 +1,431 @@
+<?php
+
+final class WP_Recovery_Mode_Email_Controller implements WP_Recovery_Mode_Controller {
```
As mentioned above, let's make this the `WP_Recovery_Mode_Controller` class and ditch the interface.

TimothyBJacobs replied:
Removed.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-controller.php:
```diff
@@ -0,0 +1,431 @@
+	private $is_active;
+
+	/** @var string|false */
+	private $session_id = false;
```
All these properties need a description and a `since` annotation.

TimothyBJacobs replied:
Added

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-controller.php:
```diff
@@ -0,0 +1,431 @@
+	 * @param WP_Recovery_Mode_Cookie_Service $cookies
+	 * @param WP_Recovery_Mode_Key_Service    $keys
+	 */
+	public function __construct( WP_Recovery_Mode_Cookie_Service $cookies, WP_Recovery_Mode_Key_Service $keys ) {
```
Similarly to ditching the interface, let's ditch dependency injection here. We can always add that back later. Let's have the constructor instantiate these classes in the properties instead.

TimothyBJacobs replied:
Removed.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-controller.php:
```diff
@@ -0,0 +1,431 @@
+	public function handle_error( array $error ) {
+		if ( is_protected_endpoint() ) {
+			$this->maybe_send_recovery_mode_email( $error );
+		}
```
As mentioned above, this should do a few more things:
* return a boolean to indicate success/failure
* take over some code from `wp_record_extension_error()` to check where the error is coming from
    * if from core/drop-in/MU-plugin/network plugin, return false
    * otherwise, try storing the error (if that fails, return false)
    * of course this doesn't all need to go in one method, could also become separate protected methods
* if on a protected endpoint and storing the error was successful, take care of redirecting

This moves responsibility for all these tasks to the recovery mode controller, rather than having that responsibility shattered across several core areas.

TimothyBJacobs replied:
Resolved. This returns `WP_Error` or `die`s I couldn't find a case where it'd return true.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-controller.php:
```diff
@@ -0,0 +1,431 @@
+		 *
+		 * @param int $rate_limit Time to wait in seconds. Defaults to 4 hours.
+		 */
+		return apply_filters( 'recovery_mode_email_rate_limit', 4 * HOUR_IN_SECONDS );
```
Just curious, why 4? We previously discussed 1 hour as the default.

TimothyBJacobs replied:
This was discussed in the Thursday meeting: https://wordpress.slack.com/archives/C60K3MP2Q/p1550761315015300

I'm fine with either.

TimothyBJacobs replied:
Leaving as 4 hours and resetting the rate limit when exiting recovery mode: https://wordpress.slack.com/archives/C60K3MP2Q/p1552928344557400

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-controller.php:
```diff
@@ -0,0 +1,431 @@
+
+		if ( ! $last_sent || time() > $last_sent + $rate_limit ) {
+			$sent = $this->send_recovery_mode_email( $error );
+			update_site_option( 'recovery_mode_email_last_sent', time() );
```
I think we should update the option before sending the email, and bail if that call was unsuccessful: If we can't store when the email was last sent for whatever internal bug, it would effectively disable the rate limit.

TimothyBJacobs replied:
Changed.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-controller.php:
```diff
@@ -0,0 +1,431 @@
+	 *
+	 * @return string
+	 */
+	private function get_error_details( $error ) {
```
Can we possibly merge some of this with code from `wp_get_plugin_error_description()`? There appears to be quite some overlap. Maybe `wp_get_plugin_error_description()` should also get a more generic name (not include "plugin").

TimothyBJacobs replied:
Merged and stripped tags in the email version.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-controller.php:
```diff
@@ -0,0 +1,431 @@
+			}
+
+			// Multiple plugins can technically be in the same directory.
+			$cause = wp_sprintf( _n( 'This may be caused by the %l plugin.', 'This may be caused by the %l plugins.', count( $names ) ), $names );
```
Even though this can technically be the case, let's not use "may" here, as it will just cause unnecessary uncertainty in most cases.

TimothyBJacobs replied:
Going to leave this for the specific [copy review ticket](https://github.com/WordPress/servehappy/issues/40).

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-controller.php:
```diff
@@ -0,0 +1,431 @@
+			$theme = wp_get_theme( $extension['slug'] );
+			$name  = $theme->exists() ? $theme->display( 'Name' ) : $extension['slug'];
+
+			$cause = sprintf( __( 'This may be caused by the %s theme.' ), $name );
```
Similarly, let's not use "may" here.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-key-service.php:
```diff
@@ -0,0 +1,84 @@
+<?php
+
+final class WP_Recovery_Mode_Key_Service {
```
This needs file and class docs.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-key-service.php:
```diff
@@ -0,0 +1,84 @@
+		global $wp_hasher;
+
+		if ( ! function_exists( 'wp_generate_password' ) ) {
+			require_once ABSPATH . WPINC . '/pluggable.php';
```
There shouldn't be file inclusions anywhere unless the same piece of code also ensures that execution stops afterwards (with `exit` or `die()`).

TimothyBJacobs replied:
Moved to `WP_Recovery_Mode::handle_error()`. I think this is the highest it could reasonably go, the next level would be `WP_Fatal_Error_Handler` itself, but I think that's overkill.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-key-service.php:
```diff
@@ -0,0 +1,84 @@
+		do_action( 'generate_recovery_mode_key', $key );
+
+		if ( empty( $wp_hasher ) ) {
+			require_once ABSPATH . WPINC . '/class-phpass.php';
```
Same here. Such file inclusions should happen in a more high level class that also ensures the process terminates after the recovery code has run. This ensures we don't accidentally miss a case and therefore run into double required file / pluggable already existing issues.

TimothyBJacobs replied:
I don't think this is necessary for `phpass`. `require_once ABSPATH . WPINC . '/class-phpass.php'` is a pattern throughout core because it is only loaded when necessary.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-key-service.php:
```diff
@@ -0,0 +1,84 @@
+		}
+
+		if ( ! function_exists( 'wp_check_password' ) ) {
+			require_once ABSPATH . WPINC . '/pluggable.php';
```
Same here.

felixarntz commented on src/wp-includes/error-protection.php:
```diff
@@ -0,0 +1,239 @@
+ *
+ * @return bool Whether the error was correctly recorded.
+ */
+function wp_record_extension_error( $error ) {
```
Let's ditch this function. The respective code should be moved to the recovery mode controller. This isn't a function anybody should easily be able to call, and not having it as a function makes that less likely.

TimothyBJacobs replied:
Removed.

felixarntz commented on src/wp-includes/error-protection.php:
```diff
@@ -0,0 +1,239 @@
+ *
+ * @return WP_Paused_Extensions_Storage
+ */
+function wp_paused_extensions() {
```
Let's split that into `wp_paused_plugins()` and `wp_paused_themes()`, similar to how it was originally.

TimothyBJacobs replied:
Resplit.

felixarntz commented on src/wp-includes/error-protection.php:
```diff
@@ -0,0 +1,239 @@
+ *                     Slug is the plugin or theme directory as opposed to the full file.
+ *                     Or false on error.
+ */
+function wp_get_extension_for_error( $error ) {
```
Similar to `wp_record_extension_error()`, this should be removed and the functionality should become part of the recovery mode controller.

TimothyBJacobs replied:
Moved.

felixarntz commented on src/wp-admin/includes/plugin.php:
```diff
@@ -2101,3 +2104,198 @@ function wp_add_privacy_policy_content( $plugin_name, $policy_text ) {
+ * @return bool|WP_Error True on success, false if `$plugin` was not paused,
+ *                       `WP_Error` on failure.
+ */
+function resume_plugin( $plugin, $redirect = '', $network_wide = false ) {
```
Let's ditch the `$network_wide` parameter.

TimothyBJacobs replied:
Removed.

felixarntz commented on src/wp-includes/error-protection.php:
```diff
@@ -0,0 +1,239 @@
+ *
+ * @return bool Whether the extension error was successfully forgotten.
+ */
+function wp_forget_extension_error( $type, $extension, $network_wide = false ) {
```
Let's ditch this function as well. We don't need a `$network_wide` parameter, and the rest of the functionality can easily accounted for by just using the paused extensions storage.

TimothyBJacobs replied:
Removed.

felixarntz commented on src/wp-includes/error-protection.php:
```diff
@@ -0,0 +1,239 @@
+ *
+ * @return bool Whether WordPress should handle this error.
+ */
+function wp_should_handle_error( $error ) {
```
This should also be ditched and moved to the recovery mode handler.

TimothyBJacobs replied:
Moved out of the global namespace. In a different class temporarilyish.

felixarntz commented on src/wp-includes/error-protection.php:
```diff
@@ -0,0 +1,239 @@
+
+		if ( defined( 'WP_CONTENT_DIR' ) && is_readable( WP_CONTENT_DIR . '/recovery-mode-controller.php' ) ) {
+			$wp_recovery_mode = include WP_CONTENT_DIR . '/recovery-mode-controller.php';
+		}
```
As mentioned above, let's not support a drop-in here.

felixarntz commented on src/wp-includes/error-protection.php:
```diff
@@ -0,0 +1,239 @@
+		 *
+		 * @param WP_Recovery_Mode_Controller $wp_recovery_mode
+		 */
+		$wp_recovery_mode = apply_filters( 'wp_recovery_mode_controller', $wp_recovery_mode );
```
I think in the first iteration we shouldn't make this overrideable at all. It can introduce significant roadblocks for the future. Let's remove for now and reconsider for v2.

felixarntz commented on tests/phpunit/tests/user/capabilities.php:
```diff
@@ -454,6 +458,7 @@ public function testMetaCapsTestsAreCorrect() {
 			// Singular object meta capabilities (where an object ID is passed) are not tested:
 			$expected['activate_plugin'],
 			$expected['deactivate_plugin'],
+			$expected['resume_plugin'],
```
Related to the suggestion to introduce `resume_theme`, that capability needs to be included here as well then.

TimothyBJacobs replied:
Added

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-controller.php:
```diff
@@ -0,0 +1,431 @@
+	/**
+	 * @inheritdoc
+	 */
+	public function run() {
```
Let's add a check for a constant here:
```php
if ( defined( 'WP_RECOVERY_MODE_SESSION_ID' ) {
    $this->is_active = true;
    $this->session_id = WP_RECOVERY_MODE_SESSION_ID;
    return;
}
```

This allows people to short-circuit everything while not tying in too much with our early class infrastructure.

TimothyBJacobs replied:
Added.

felixarntz reviewed (changes requested):
I did a more thorough review and pointed out some details. Most of it is looking good.

A few more general things:
* Let's keep the `WP_Fatal_Error_Handler` class as generic as it can be. Storing errors, redirecting, etc. already ties in with recovery mode and how we envision it, so these things should be covered by the recovery mode controller. The only connection should be the fatal error handler calling the `handle_fatal_error()` method.
* Let's be conservative and not expose too much code through functions that we only call in one location. We use class structures here, and the feature is in a first iteration, so we don't want to expose too much for outside usage here.
* The same applies to overriding. People can override the fatal error handler, and thus technically already have the ability to override in whichever way they want. For the first iteration, let's not get more granular, i.e. not allow to override the recovery mode handler specifically. In `is_recovery_mode_active()`, we can check a constant that MU plugins can define - that allows for customizing that bit, while not restricting us in what we can do to our class infrastructure later. We don't want to corner ourselves.
* Function/method descriptions in doc-blocks should start with a third-person verb rather than infinitive. :)

TimothyBJacobs committed:
Remove multisite/network admin handling.

TimothyBJacobs committed:
Update paused theme capability checks.

TimothyBJacobs committed:
Update test for resume_theme meta cap.

TimothyBJacobs committed:
Update extension storage to restore separate object instances,
but save the error to the same option. Drop multisite option handling.
This can be reintroduced safely in a later version when we want to
support Multisite.

felixarntz commented:
Just taking a note here that we should wrap the calls in `wp-settings.php` into a `! is_multisite()`, as the feature will be disabled for multisite in v1.

TimothyBJacobs committed:
Drop recovery mode controller interface and drop-in for v1.

Now `WP_Recovery_Mode` is the object responsible for managing recovery mode.
WP_Fatal_Error_Handler passes off control to `WP_Recovery_Mode::handle_error()`
when a fatal error occurs.

`WP_Recovery_Mode_Email_Controller` is renamed to `WP_Recovery_Mode_Email`.
This keeps the line between general Recovery Mode processing, like
handling storing errors and doing redirects, separate from the method
of notifying the user to enter recovery mode.

Adjusted redirect_protected functionality to account for when headers
have already been sent ( display errors being on ). This right now just
displays the default error template. We might want to adjust that.

Adds a admin menu bar link to exit recovery mode.

The cookie check can be overridden by using the `WP_RECOVERY_MODE_SESSION_ID`
constant, but this probably shouldn't be encouraged.

TimothyBJacobs committed:
Don't initialize recovery mode in multisite.

felixarntz commented on src/wp-includes/class-wp-recovery-mode.php:
```diff
@@ -0,0 +1,353 @@
+	 */
+	public function __construct() {
+		$this->email   = new WP_Recovery_Mode_Email_Link();
+		$this->cookies = new WP_Recovery_Mode_Cookie_Service();
```
The fact you instantiate `WP_Recovery_Mode_Cookie_Service` both here and in `WP_Recovery_Mode_Email_Link` tells me there's something wrong.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-link.php:
```diff
@@ -0,0 +1,348 @@
+ *
+ * @since 5.2.0
+ */
+final class WP_Recovery_Mode_Email_Link {
```
This class should be simplified to only be responsible for the actual email sending. The recovery link handling is agnostic of how the user obtains that link. I think it should be moved to `WP_Recovery_Mode` or a new `WP_Recovery_Mode_Link_Service`.

TimothyBJacobs replied:
Extracted to a separate link service.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-link.php:
```diff
@@ -0,0 +1,348 @@
+	 */
+	public function __construct() {
+		$this->keys    = new WP_Recovery_Mode_Key_Service();
+		$this->cookies = new WP_Recovery_Mode_Cookie_Service();
```
This class shouldn't need to worry about the recovery cookie.

TimothyBJacobs replied:
Removed both services.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-link.php:
```diff
@@ -0,0 +1,348 @@
+	 *
+	 * @since 5.2.0
+	 */
+	public function handle_begin_link() {
```
This method should be moved elsewhere.

TimothyBJacobs replied:
Moved.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-link.php:
```diff
@@ -0,0 +1,348 @@
+	 *
+	 * @return string
+	 */
+	private function get_recovery_mode_begin_url( $key ) {
```
This method should be moved elsewhere.

TimothyBJacobs replied:
Moved.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-link.php:
```diff
@@ -0,0 +1,348 @@
+	 *
+	 * @return int Interval in seconds.
+	 */
+	private function get_link_valid_for_interval() {
```
This method should be moved elsewhere.

TimothyBJacobs replied:
Moved.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-link.php:
```diff
@@ -0,0 +1,348 @@
+			$theme = wp_get_theme( $extension['slug'] );
+			$name  = $theme->exists() ? $theme->display( 'Name' ) : $extension['slug'];
+
+			$cause = sprintf( __( 'This may be caused by the %s theme.' ), $name );
```
As mentioned before, I'd refrain from using "may" here. Let's just say "This was caused by ...".

TimothyBJacobs replied:
Updated.

felixarntz commented on src/wp-includes/error-protection.php:
```diff
@@ -0,0 +1,137 @@
+	}
+
+	/* translators: 1: error type, 2: error line number, 3: error file name, 4: error message */
+	$error_message = __( 'The plugin caused an error of type %1$s in line %2$s of the file %3$s. Error message: %4$s' );
```
This is no longer plugin-specific. Let's rephrase to "An error of type %1$s was caused in line %2$s of the file %3$s."

In the areas where this function is used, it's obvious by the context which plugin or theme caused it anyway.

TimothyBJacobs replied:
Updated.

felixarntz commented on src/wp-includes/class-wp-fatal-error-handler.php:
```diff
@@ -0,0 +1,172 @@
+				return;
+			}
+
+			wp_recovery_mode()->handle_error( $error );
```
Let's wrap this in a `! is_multisite()` clause, just to be sure.

TimothyBJacobs replied:
Added

felixarntz commented on src/wp-settings.php:
```diff
@@ -17,9 +17,15 @@
 require_once( ABSPATH . WPINC . '/plugin.php' );
 
+// Make sure we register the shutdown handler for fatal errors as soon as possible.
+wp_register_fatal_error_handler();
```
We need to discuss whether this should also be excluded if `! is_multisite()`.

spacedmonkey replied:
I think we can support that feature in multisite.

TimothyBJacobs replied:
Looks like we decided to not exclude it: https://wordpress.slack.com/archives/C60K3MP2Q/p1553074572612600

felixarntz commented on src/wp-settings.php:
```diff
@@ -261,6 +267,10 @@
+require( ABSPATH . WPINC . '/class-wp-recovery-mode-cookie-service.php' );
+require( ABSPATH . WPINC . '/class-wp-recovery-mode-key-service.php' );
+require( ABSPATH . WPINC . '/class-wp-recovery-mode-email-link.php' );
+require( ABSPATH . WPINC . '/class-wp-recovery-mode.php' );
```
Looking at these classes and the other observations, I suggest we do the following:
* Introduce `WP_Recovery_Mode_Link_Service`, moving most code that is currently in `WP_Recovery_Mode_Email_Link` there. The `WP_Recovery_Mode_Key_Service` instance should also be moved there, as the email handler doesn't need to know about it. The `WP_Recovery_Mode_Email_Link` instance should be controlled by `WP_Recovery_Mode`.
* Rename `WP_Recovery_Mode_Email_Link` to `WP_Recovery_Mode_Email_Service`.
* We need to allow `WP_Recovery_Mode_Email_Service` to get a recovery link, for inclusion in the email. I suggest we add a utility function `wp_get_recovery_mode_link()` that wraps the respective method in `WP_Recovery_Mode_Link_Service` (accessed via `wp_recovery_mode()`). The email class can then call this function. For that part I'd argue a wrapper function is useful so that developers can easily work with recovery links, e.g. for implementing an alternative way to retrieve it. The alternative to that function would be dependency injection, but as mentioned before I'd rather not add constructor parameters in this early stage of the API.

TimothyBJacobs replied:
I followed that path mostly, but passed the link service as a method parameter to avoid introducing a global function that back references the recovery mode class that is calling the function in the first place.

felixarntz replied:
As mentioned, I'm okay with passing it, but we should move that to the constructor to not pollute method signatures.

felixarntz replied:
As mentioned, I'm okay with passing it, but we should move that to the constructor to not pollute method signatures.

TimothyBJacobs replied:
Will adjust.

felixarntz commented on src/wp-settings.php:
```diff
@@ -340,6 +350,11 @@
 register_theme_directory( get_theme_root() );
 
+if ( ! is_multisite() ) {
+// Handle users requesting a recovery mode link and initiating recovery mode.
```
Wrong indentation :)

TimothyBJacobs replied:
Fixed.

felixarntz commented on tests/phpunit/tests/recovery-mode.php:
```diff
@@ -0,0 +1,151 @@
+<?php
+
+class Tests_Recovery_Mode extends WP_UnitTestCase {
```
Can you add `@group error-protection` to the class doc?

TimothyBJacobs replied:
Added

felixarntz commented on tests/phpunit/tests/recovery-mode.php:
```diff
@@ -0,0 +1,151 @@
+		return parent::tearDownAfterClass();
+	}
+
+	public function test_generate_and_store_returns_recovery_key() {
```
Can you add `@ticket 46130` to all test method annotations?

TimothyBJacobs replied:
Added

felixarntz reviewed (changes requested):
This is getting close. A few outstanding comments.

TimothyBJacobs committed:
Some more refactoring to separate the generation and validation of links
from notifying the user.

spacedmonkey commented on src/wp-includes/class-wp-fatal-error-handler.php:
```diff
@@ -0,0 +1,174 @@
+				return;
+			}
+
+			if ( ! is_multisite() ) {
```
Why is this disabled on multisite?

TimothyBJacobs replied:
From the summary of the meeting in slack: https://wordpress.slack.com/archives/C60K3MP2Q/p1552927814544600

> We're not going to support multisite in v1.

spacedmonkey replied:
To be discussed

TimothyBJacobs replied:
Looks like we are going to keep the fatal error handler on Multisite, but not enter the recovery mode controller: https://wordpress.slack.com/archives/C60K3MP2Q/p1553074572612600

spacedmonkey commented on src/wp-includes/class-wp-fatal-error-handler.php:
```diff
@@ -0,0 +1,174 @@
+	 */
+	public function handle() {
+		// Bail if WordPress executed successfully.
+		if ( defined( 'WP_EXECUTION_SUCCEEDED' ) && WP_EXECUTION_SUCCEEDED ) {
```
I have posted my thoughts on this [here](https://core.trac.wordpress.org/ticket/45989#comment:25). I don't like the `WP_EXECUTION_SUCCEEDED` constant.

felixarntz replied:
Agreed. Let's continue discussion for that on the ticket though, it can go in after the first commit.

spacedmonkey replied:
@felixarntz why can't we fix this now?

spacedmonkey commented on src/wp-includes/class-wp-fatal-error-handler.php:
```diff
@@ -0,0 +1,174 @@
+			return false;
+		}
+
+		$error_types_to_handle = array(
```
Should this have a filter?

TimothyBJacobs replied:
I think this is worthy of a filter. What say you @felixarntz.

felixarntz replied:
SGTM, generally. Again though, we should probably apply the filter in a way that developers cannot remove error types from the list, only add further ones.

spacedmonkey replied:
I disagree, there maybe good you reason you to want to remove something from this list. If you want to remove something here, then that is up to you as a developer to decide.

spacedmonkey commented on src/wp-includes/class-wp-fatal-error-handler.php:
```diff
@@ -0,0 +1,174 @@
+		 */
+		$args = apply_filters( 'wp_php_error_args', $args );
+
+		wp_die( $message, '', $args );
```
Should this pass a `WP_Error` here? So the code in `wp_die` json handler shows a code, so easier to find where the error comes from.

TimothyBJacobs replied:
That makes sense to me. Adjusted to a WP_Error with code `internal_server_error`.

spacedmonkey commented on src/wp-includes/class-wp-paused-extensions-storage.php:
```diff
@@ -0,0 +1,182 @@
+			return false;
+		}
+
+		$paused_extensions = (array) get_option( $this->option_name, array() );
```
Maybe we could use the function `__get_option` here. `get_option` function has a number of filters on it, that could break or change how this received from the database. `__get_option` is a clean function that just calls the database , aviods caches and filters.

spacedmonkey replied:
Long term do we want to store this data in blog meta. Should we do this now, as it maybe hard to change later.

spacedmonkey replied:
Should we use `register_setting` so we can see this setting in the api and setup the default.

TimothyBJacobs replied:
> Maybe we could use the function __get_option here.

I'm not sure whether `__get_option` makes sense or not. Currently it's private to `upgrade.php`. Not sure what the ramifications are of using it here.

> Long term do we want to store this data in blog meta. 

Yes, I think we'd want to use sitemeta for multisite. That was what the original extension storage used. But since we aren't supporting multisite at all the moment, we shouldn't run into any upgrade issues when we add it in 5.3.

> Should we use register_setting so we can see this setting in the api and setup the default.

The option name is dynamic so I don't think it'd be helpful. If we wanted to expose this in the REST API, I think we'd need a different strategy. Not sure how useful it'd be in a REST context though without plugin/theme management APIs.

felixarntz replied:
We can easily migrate to site meta when we start supporting multisite.

I also think it's best to keep the regular `get_option()` rather than including a private function from `upgrade.php`.

Registration with the REST API is something to consider, but I'd argue since the option name is based on a session ID, that would better be handled in an entirely separate controller. Furthermore it wouldn't provide significant benefits for now, it's not crucial for the feature.

spacedmonkey replied:
Is there any worry here that the option might be filtered and mess things up?

TimothyBJacobs replied:
I'm less worried because I think the filters would only be available to `mu-plugins` or network activated plugins which we aren't pausing anyways.

spacedmonkey commented on src/wp-includes/class-wp-recovery-mode-email-service.php:
```diff
@@ -0,0 +1,223 @@
+ *
+ * @since 5.2.0
+ */
+final class WP_Recovery_Mode_Email_Service {
```
Why is does this need to be final?

TimothyBJacobs replied:
I think for 5.2 we want to keep the API as closed as possible so we are more free to make changes in 5.3

spacedmonkey replied:
Okay, I guess it can always be changed later.

spacedmonkey commented on src/wp-includes/class-wp-recovery-mode-email-service.php:
```diff
@@ -0,0 +1,223 @@
+	 *
+	 * @return true|WP_Error True if email sent, WP_Error otherwise.
+	 */
+	public function maybe_send_recovery_mode_email( WP_Recovery_Mode_Link_Service $link_service, $rate_limit, $error, $extension ) {
```
Is this supported in 5.2? `WP_Recovery_Mode_Link_Service $link_service`

TimothyBJacobs replied:
Yes: https://3v4l.org/SiKBC

spacedmonkey commented on src/wp-includes/class-wp-recovery-mode-email-service.php:
```diff
@@ -0,0 +1,223 @@
+
+		$last_sent = get_site_option( self::RATE_LIMIT_OPTION );
+
+		if ( ! $last_sent || time() > $last_sent + $rate_limit ) {
```
Other parts of core use transients as a lock here. See `_get_cron_lock`. Which is not a bad pattern. It would be self clearing in the database.

TimothyBJacobs replied:
I'm not necessarily opposed to a transient here. We can't rely on the transient to be there for the entire TTL though. Is that a problem? It'd also limit the utility of the rate limit being filterable since it'd have to wait at least a cycle to become effective ( we can't change the TTL after the transient has already been set ).

felixarntz replied:
I agree we shouldn't use `get_site_option()` (i.e. network options) here. Switching to a transient has the problem that we can't rely on its existence though, as @TimothyBJacobs says. This is something we shouldn't consider for something as critical as a rate limit. I suggest using regular option (`get_option()` and `update_option()`).

spacedmonkey replied:
I am okay with `get_option`

TimothyBJacobs replied:
Updated to regular option functions.

spacedmonkey commented on src/wp-includes/class-wp-recovery-mode-email-service.php:
```diff
@@ -0,0 +1,223 @@
+	private function send_recovery_mode_email( WP_Recovery_Mode_Link_Service $link_service, $rate_limit, $error, $extension ) {
+
+		$url = $link_service->generate_url();
+		$blogname = wp_specialchars_decode( get_option( 'blogname' ), ENT_QUOTES );
```
Shouldn't we use `get_bloginfo` here?

TimothyBJacobs replied:
I based this on `send_confirmation_on_profile_email()`, `update_option_new_admin_email()`, `wp_send_user_request()` and other similar functions which use this pattern. I'm really not sure why they do it that way.

spacedmonkey commented on src/wp-includes/class-wp-recovery-mode-key-service.php:
```diff
@@ -0,0 +1,87 @@
+
+		$hashed = $wp_hasher->HashPassword( $key );
+
+		update_site_option( 'recovery_key', array(
```
Why would you want this stored at network level if multisite is enabled.

TimothyBJacobs replied:
As opposed to? This matches the behavior of `wp_salt()`.

TimothyBJacobs replied:
Sorry, got this confused with the other comment. I'm not sure what the correct behavior is here for when we'd support multisite. Atm we could just change this to the `get_option` APIs. @felixarntz

TimothyBJacobs replied:
_ignore this ui keeps jumping around on me_

felixarntz replied:
Let's refrain from using network options, and always use regular options at this point - regardless of `wp_salt()` does it. Using a network option here would potentially conflict with our solution for multisite later. Let's switch to `update_option()` / `get_option()`.

TimothyBJacobs replied:
Updated.

spacedmonkey commented on src/wp-admin/plugins.php:
```diff
@@ -389,6 +389,26 @@
 			break;
 
+		case 'resume':
+			if ( is_multisite() && is_network_admin() ) {
```
Why is this disabled?

TimothyBJacobs replied:
We want to make sure we aren't accidentally running any code in a multisite context. It should probably be loosened to just `is_multisite()`. When we introduce multisite support, we'd remove these, but I think we want to be as safe as possible atm.

spacedmonkey commented on src/wp-includes/admin-bar.php:
```diff
@@ -1043,6 +1043,36 @@ function wp_admin_bar_search_menu( $wp_admin_bar ) {
+			'title'  => __( 'Exit Recovery Mode' ),
+			'href'   => $url,
+			'meta'   => array(
+				'title' => __( 'Exit Recovery Mode' )
```
Can we better explain what this button does somehow? Can we highlight this button

TimothyBJacobs replied:
Yeah, this needs exploration on the design side.

spacedmonkey replied:
I would love it if this admin bar button was massive a red so you can't miss it.

TimothyBJacobs replied:
Going to resolve this for now. I believe @felixarntz is involving the design team.

spacedmonkey commented on src/wp-includes/class-wp-fatal-error-handler.php:
```diff
@@ -0,0 +1,174 @@
+			$php_error_pluggable = WP_CONTENT_DIR . '/php-error.php';
+			if ( is_readable( $php_error_pluggable ) ) {
+				require_once $php_error_pluggable;
+
```
Should this work like `db-error.php`, this calls `die()` after the file is included.

TimothyBJacobs replied:
That caused a compatibility issue with Query Monitor and other plugins that have their own shutdown hooks. https://core.trac.wordpress.org/ticket/46038

spacedmonkey commented on src/wp-includes/class-wp-recovery-mode-cookie-service.php:
```diff
@@ -0,0 +1,281 @@
+	public function get_session_id_from_cookie( $cookie = '' ) {
+		if ( ! $cookie ) {
+			if ( empty( $_COOKIE[ $this->name ] ) ) {
+				return new WP_Error( 'no_cookie' );
```
Should this have a message?

TimothyBJacobs replied:
Good catch, fixed.

spacedmonkey commented on src/wp-includes/class-wp-recovery-mode-cookie-service.php:
```diff
@@ -0,0 +1,281 @@
+	private function recovery_mode_hash( $data ) {
+
+		if ( ! defined( 'AUTH_KEY' ) || 'put your unique phrase here' === AUTH_KEY ) {
+			$auth_key = get_site_option( 'recovery_mode_auth_key' );
```
Why is this stored at network level?

felixarntz replied:
I think this is the only place where a network option is fine. That is because this is essentially a "bad" fallback for when the constant is not properly configured. The constant is global for an entire (multi-)site, therefore I think here we can keep using network option.

spacedmonkey commented on src/wp-includes/class-wp-recovery-mode-cookie-service.php:
```diff
@@ -0,0 +1,281 @@
+		}
+
+		if ( ! defined( 'AUTH_SALT' ) || 'put your unique phrase here' === AUTH_SALT || $auth_key === AUTH_SALT ) {
+			$auth_salt = get_site_option( 'recovery_mode_auth_salt' );
```
Why is this stored at network level?

TimothyBJacobs replied:
As opposed to? This matches the behavior of wp_salt().

felixarntz replied:
Same here, this should keep using network options since it replaces a constant (which by definition is global).

spacedmonkey replied:
If it is a constant, then it could conditional by domain.

spacedmonkey commented on src/wp-includes/class-wp-recovery-mode.php:
```diff
@@ -0,0 +1,403 @@
+	 */
+	protected function redirect_protected() {
+		// Pluggable is usually loaded after plugins, so we manually include it here for redirection functionality.
+		if ( ! function_exists( 'wp_redirect' ) ) {
```
Check for `wp_safe_redirect`

TimothyBJacobs replied:
Updated.

spacedmonkey commented on src/wp-includes/error-protection.php:
```diff
@@ -0,0 +1,137 @@
+ * @return bool True if the fatal error handler is enabled, false otherwise.
+ */
+function wp_is_fatal_error_handler_enabled() {
+	$enabled = ! defined( 'WP_DISABLE_FATAL_ERROR_HANDLER' ) || ! WP_DISABLE_FATAL_ERROR_HANDLER;
```
Love this :+1:

TimothyBJacobs replied:
Credit goes to @felixarntz for this I believe.

spacedmonkey commented on src/wp-includes/load.php:
```diff
@@ -725,9 +762,146 @@ function wp_get_active_and_valid_themes() {
+ * @return bool True if the current endpoint should be protected.
+ */
+function is_protected_endpoint() {
+	// Protect login pages.
```
Return in this function makes the filter useless here.
It should be more like this 
```
$protected = false;
if ( isset( $GLOBALS['pagenow'] ) && 'wp-login.php' === $GLOBALS['pagenow'] ) {
	$protected = true;
}
return (bool) apply_filters( 'is_protected_endpoint', $protected );

TimothyBJacobs replied:
That's intentional. From the hook doc:

```
This filter is only fired when an endpoint is requested which is not already protected WordPress core. As such, it exclusively allows providing further protected endpoints in addition to the admin backend, login pages and protected AJAX actions.
```

And from @felixarntz original patch comments: https://core.trac.wordpress.org/ticket/44458#comment:34

spacedmonkey replied:
Yeah, I didn't agree with that the first time around. Having conditional filter is so confusing for developers.

felixarntz replied:
It's important to ensure that developers can only add further endpoints, not remove existing ones. The way this code is written covers that.

spacedmonkey replied:
Okay, then the filter should run before all other logic, like other short circuits

spacedmonkey commented on src/wp-settings.php:
```diff
@@ -340,6 +351,11 @@
 // Register the default theme directory root
 register_theme_directory( get_theme_root() );
 
+if ( ! is_multisite() ) {
```
We need to discuss this

TimothyBJacobs replied:
Going to resolve this in favor of this comment, https://github.com/wp-core-php/wordpress-develop/pull/6#discussion_r267036435 since it appears first :)

spacedmonkey reviewed (changes requested):
That is my code review for now.

TimothyBJacobs committed:
Pass a WP_Error object to wp_die call so the error code can be seen
by non HTML handlers.

TimothyBJacobs committed:
Add message to WP_Error when no cookie. This was fixed previously for the
validate_cookie method but not get_session_id_from_cookie.

TimothyBJacobs committed:
Be more specific in our function check for loading pluggable.php

spacedmonkey commented:
@TimothyBJacobs All references to *_site_option should be changed  *_option, so we can better support multisite in the future.

felixarntz commented:
> All references to *_site_option should be changed *_option, so we can better support multisite in the future.

I agree, with the exception of those that are a fallback for a constant not being properly defined, since those are by definition global. As far as I can tell, those options would only be populated and used under these circumstances anyway.

felixarntz commented on src/wp-includes/class-wp-paused-extensions-storage.php:
```diff
@@ -0,0 +1,182 @@
+	 */
+	public function __construct( $extension_type ) {
+		$this->type = $extension_type;
+		$this->option_name = wp_recovery_mode()->get_session_id() . '_paused_extensions';
```
These two equal signs should be aligned, per WPCS.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-service.php:
```diff
@@ -0,0 +1,223 @@
+	 *
+	 * @return true|WP_Error True if email sent, WP_Error otherwise.
+	 */
+	public function maybe_send_recovery_mode_email( WP_Recovery_Mode_Link_Service $link_service, $rate_limit, $error, $extension ) {
```
Given that we decided against introducing a global function for the recovery link URL, I'd argue under these circumstances it _is_ better to use dependency injection with the constructor of the class, rather than individual methods. Passing instances to methods directly locks us in more than with the constructor I'd argue.
Can you move the `WP_Recovery_Mode_Link_Service` instance to a property, passed into the constructor?

Sorry about the back and forth here.

TimothyBJacobs replied:
Moved.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-service.php:
```diff
@@ -0,0 +1,223 @@
+	 * @return bool
+	 */
+	public function clear_rate_limit() {
+		return delete_site_option( self::RATE_LIMIT_OPTION );
```
This should use `delete_option()`, following our other discussion.

TimothyBJacobs replied:
Updated.

felixarntz commented on src/wp-includes/class-wp-recovery-mode-email-service.php:
```diff
@@ -0,0 +1,223 @@
+			}
+
+			// Multiple plugins can technically be in the same directory.
+			$cause = wp_sprintf( _n( 'This was caused by the %l plugin.', 'This was be caused by the %l plugins.', count( $names ) ), $names );
```
The clause for multiple plugins would read weird to a user. We should:
* either ditch this, only use the first name if there are multiple (I think that's perfectly fine because the rest is quite the edge-case)
* or "over-complicate" the code to ensure this is a nice comma-separated list with the last two items separated by "and"

I'd argue 1 is cleaner and covers 99% of use-cases, so I'd vote for that. But I leave the decision to you.

TimothyBJacobs replied:
I think it is a comma separated list. It's using `wp_sprintf` and the `%l` format which should return a comma separated last is "and" string.

TimothyBJacobs replied:
Yeah. It currently reads like

```
"This was be caused by the Test A and Test B plugins."
```

```
"This was be caused by the Test A, Test B, Test C, and Test D plugins."
```

felixarntz commented on src/wp-includes/class-wp-recovery-mode.php:
```diff
@@ -0,0 +1,403 @@
+	public function __construct() {
+		$this->email        = new WP_Recovery_Mode_Email_Service();
+		$this->link_handler = new WP_Recovery_Mode_Link_Service();
+		$this->cookies      = new WP_Recovery_Mode_Cookie_Service();
```
These properties should probably be named in a unified way: Either all should have a `_service` suffix, or none should have a suffix. I'd refrain from using `_handler` since the class is a service.

TimothyBJacobs replied:
Uses `_service` suffix.

felixarntz commented on src/wp-includes/class-wp-paused-extensions-storage.php:
```diff
@@ -0,0 +1,182 @@
+	 */
+	public function __construct( $extension_type ) {
+		$this->type = $extension_type;
+		$this->option_name = wp_recovery_mode()->get_session_id() . '_paused_extensions';
```
I know the way the code is currently written, it can't happen, but we should cover here for the case that no session ID is available. Theoretically anyone could call `wp_paused_plugins()`/`wp_paused_themes()` at any time, potentially instantiating the class. We can't rely on the recovery mode cookie being read and the session ID being available on instantiation of this class.

To address these concerns, I suggest the following:
* We don't store the option name in a property.
* We introduce a private method `get_option_name()`:
    * It calls `wp_recovery_mode()->get_session_id()`.
    * If empty, the function returns an empty string.
    * If not empty, it returns `$session_id . '_paused_extensions'`.
* Every method that tries to access the option calls the above method. This ensures we always request the session ID when we need it rather than potentially doing it too early. Methods that use the name multiple times could store it in a variable once to make the code more readable. This shouldn't have performance implications since the session ID is elsewhere already stored in a property, so that no expensive logic needs to run.
* If the option name retrieved from the method is empty, each method should return false (or its respective default value) rather than doing anything.

TimothyBJacobs replied:
Updated.

felixarntz reviewed (changes requested):
Looks good, now it's getting really granular with concerns. :)

I will run some testing today and tomorrow, otherwise I'd say once the outstanding feedback is addressed, this is good for commit.

TimothyBJacobs committed:
Switch away form site option APIs where applicable.

TimothyBJacobs committed:
Loosen multisite check to just check multisite at all, not also the network admin.

spacedmonkey commented on src/wp-admin/includes/plugin.php:
```diff
@@ -2101,3 +2103,134 @@ function wp_add_privacy_policy_content( $plugin_name, $policy_text ) {
+		sprintf(
+			'<a href="%s">%s</a>',
+			admin_url( 'plugins.php?plugin_status=paused' ),
+			'Go to the Plugins screen'
```
Missed translation.

TimothyBJacobs replied:
Translated.

spacedmonkey commented on src/wp-admin/includes/plugin.php:
```diff
@@ -2101,3 +2103,134 @@ function wp_add_privacy_policy_content( $plugin_name, $policy_text ) {
+		'<div class="notice notice-error"><p><strong>%s</strong><br>%s</p><p>%s</p></div>',
+		__( 'One or more plugins failed to load properly.' ),
+		__( 'You can find more details and make changes on the Plugins screen.' ),
+		sprintf(
```
Why is this another sprintf, this is confusing. Just use one.

TimothyBJacobs replied:
Consolidated

spacedmonkey commented on src/wp-admin/includes/theme.php:
```diff
@@ -768,3 +768,103 @@ function customize_themes_print_templates() {
+		sprintf(
+			'<a href="%s">%s</a>',
+			admin_url( 'themes.php' ),
+			'Go to the Themes screen'
```
Missed translation.

TimothyBJacobs replied:
Translated.

TimothyBJacobs committed:
Use option name from a function return and ensure that we are in recovery
mode and have an active session ID before use.

This blocks mistakes from 3rd party code loading the storage
instance too early.

TimothyBJacobs committed:
Inject in the constructor instead of passing as method dependencies.

Consistently name properties.

TimothyBJacobs committed:
Add missing translations.

TimothyBJacobs committed:
Consolidate to a single printf call. Escape URL.

johnbillion commented on src/wp-includes/class-wp-recovery-mode-email-service.php:
```diff
@@ -0,0 +1,236 @@
+			),
+			array(
+				$url,
+				'TBD',
```
TBD!

TimothyBJacobs replied:
Is pending this issue: https://github.com/WordPress/servehappy/issues/40 :)

johnbillion commented on src/wp-includes/class-wp-recovery-mode-email-service.php:
```diff
@@ -0,0 +1,236 @@
+
+		$email = array(
+			'to'      => $this->get_recovery_mode_email_address(),
+			'subject' => __( '[%s] Your Site Experienced an Issue' ),
```
I'd like to see this using present tense and being more specific, eg `Your Site is Experiencing a Technical Issue`

TimothyBJacobs replied:
That sounds better to me as well, updated! Thanks!

johnbillion commented on src/wp-login.php:
```diff
@@ -1029,6 +1029,8 @@ function retrieve_password() {
 			} elseif ( strpos( $redirect_to, 'about.php?updated' ) ) {
 				$errors->add( 'updated', __( '<strong>You have successfully updated WordPress!</strong> Please log back in to see what&#8217;s new.' ), 'message' );
+			} elseif ( WP_Recovery_Mode_Link_Service::LOGIN_ACTION_ENTERED === $action ) {
+				$errors->add( 'enter_recovery_mode', __( 'Recovery Mode Initialized. Please login to continue.' ), 'message' );
```
```suggestion
				$errors->add( 'enter_recovery_mode', __( 'Recovery Mode Initialized. Please log in to continue.' ), 'message' );
```

johnbillion committed:
Use proper form of "log in" from @johnbillion

Co-Authored-By: TimothyBJacobs <timothy@ironbounddesigns.com>

TimothyBJacobs committed:
Update subject with johnbillion's wording to be in the present tense,
and more specific.

Commit:
Merge branch 'master' into wsod-protection-refined

felixarntz commented:
Merged against latest trunk, where the basic fatal error handler is already committed now.

Commit:
Merge branch 'master' into wsod-protection-refined

Commit:
Fix coding standards violations.

Commit:
Load recovery mode classes early in case they're accessed early by the fatal error handler.

Commit:
Minor readability tweak.

Commit:
Ensure  is a string and add some missing docs.

TimothyBJacobs committed:
Check whether recovery mode has been initialized before handling
the error.

Commit:
Add/correct some missing docs.

Commit:
Merge branch 'wsod-protection-refined' of github.com:wp-core-php/wordpress-develop into wsod-protection-refined

Commit:
Minor doc fix.

Commit:
Fix bug where redirect when resuming a still broken plugin would be slightly incorrect.

Commit:
Check whether a theme can be resumed (it should no longer cause fatal errors) and provide according feedback.

Commit:
Fix doc.

TimothyBJacobs committed:
Increase recovery mode key length to 22 characters.

Commit:
Split tests into separate files per class, and fix test failures in multisite.

Commit:
Fix capability tests for new user_has_cap capabilities.

Commit:
Merge branch 'master' into wsod-protection-refined

felixarntz commented:
Done! https://core.trac.wordpress.org/changeset/44973

Thanks for this truly outstanding work @TimothyBJacobs! 🎉

TimothyBJacobs commented:
Thank you! The recovery mode code wouldn't be anywhere near where it is without your extensive help.
